### PR TITLE
Some minor docker improvements

### PR DESCRIPTION
**Pull request by g2flyer**
A few minor improvements in docker code to
- docker & docker-compose consistent
  -  consistent tagging
  -  build only via docker to prevent double builds
  - consistent support of ADD_APT_PKGS
- some additional docker related make targets, e.g., additional stop targets, test without complete rebuild and _dev targets to setup development containers
- some minor dockerfile changes to be more consistent

BTW: in readme i've noticed we suggest in our examples to put user-name into container names. This seems a wise idea for multi-user settings. Should we also adopt it in the makefiles?  Easy to do but haven't changed it yet as i'm not sure if some other places count on the container names being what they are ...
Along the lines, the might include in name in container (and, more importantly, image) also SGX-mode (where also hw-mode in testing seems to have gone lost but see also PR#463, interpreter, ledger, ...?

### CODE REVIEW

File: docker/pdo_client.dockerfile [hunk 1]
```diff
@@ -37,6 +37,9 @@ USER $UNAME
 # -----------------------------------------------------------------
 ARG REBUILD 0
 
+ARG PDO_SOURCE_ROOT=/project/pdo/src
+ENV PDO_SOURCE_ROOT=${PDO_SOURCE_ROOT}
+
```
Comment (cmickeyb) [inline docker/pdo_client.dockerfile:42]: suggest we drop this. it should not be configurable.

File: docker/pdo_client.dockerfile [hunk 2]
```diff
@@ -54,7 +57,7 @@ ENV PDO_LOG_LEVEL=${PDO_LOG_LEVEL}
 
 # copy the source files into the image
 WORKDIR /project/pdo
-COPY --chown=${UNAME}:${UNAME} repository /project/pdo/src
+COPY --chown=${UNAME}:${UNAME} repository ${PDO_SOURCE_ROOT}
```
Comment (cmickeyb) [inline docker/pdo_client.dockerfile:61]: same. there is no reason to make pdo_source_root configurable.

File: docker/pdo_client.dockerfile [hunk 3]
```diff
@@ -66,10 +69,13 @@ COPY --chown=${UNAME}:${UNAME} tools/*.sh ./
...
 ARG PDO_HOSTNAME
-ENV PDO_HOSTNAME=$PDO_HOSTNAME
+ENV PDO_HOSTNAME=${PDO_HOSTNAME}
 
 ARG PDO_LEDGER_URL
-ENV PDO_LEDGER_URL=$PDO_LEDGER_URL
+ENV PDO_LEDGER_URL=${PDO_LEDGER_URL}
+
+ARG PDO_LEDGER_KEY_ROOT=/project/pdo/xfer/ccf/keys/
+ENV PDO_LEDGER_KEY_ROOT=${PDO_LEDGER_KEY_ROOT}
```
Comment (cmickeyb) [inline docker/pdo_client.dockerfile:79]: same. we control the layout of the container file system. we choose the structure.

**Review by cmickeyb — CHANGES_REQUESTED** [review]
Overall comment: please be consistent with "{}" or "()" for variable expansion in the Makefile. I don't have a strong feeling either way, but I would strongly prefer to go with one form or the other.

File: docker/pdo_client.dockerfile (hunk 1 above)
Comment (g2flyer) [reply docker/pdo_client.dockerfile:42]: I didn't realize that one is supposed to run start_client which then all sets this (and below).  So definitely will revert and see whether maybe we can also do the start_client stuff automagically ..

File: docker/pdo_client.dockerfile (hunk 2 above)
Comment (g2flyer) [reply docker/pdo_client.dockerfile:61]: see aboe

File: docker/pdo_client.dockerfile (hunk 3 above)
Comment (g2flyer) [reply docker/pdo_client.dockerfile:79]: see above

**Comment by g2flyer** [comment]
@cmickeyb Mic, i reworked the PR after our discussion, the PR description is now updated to the current changes (which essentially should address your comments as mentioned above)

File: docker/pdo_client.dockerfile [hunk 4]
```diff
@@ -65,13 +65,13 @@ COPY --chown=${UNAME}:${UNAME} tools/*.sh ./
 # build it!!!
 RUN /project/pdo/tools/build_client.sh
 
-ARG PDO_HOSTNAME
-ENV PDO_HOSTNAME=$PDO_HOSTNAME
+ARG PDO_HOSTNAME=localhost
+ENV PDO_HOSTNAME=${PDO_HOSTNAME}
 
-ARG PDO_LEDGER_URL
-ENV PDO_LEDGER_URL=$PDO_LEDGER_URL
```
Comment (bvavala) [inline docker/pdo_client.dockerfile:72]: Thinking about this, it seems appropriate not to include a ledger url in the image, but I am wondering if even the hostname should be provided at a later time.

File: docker/pdo_services.dockerfile [hunk 5]
```diff
@@ -55,12 +55,9 @@ EXPOSE 7001 7002 7003 7004 7005
 EXPOSE 7101 7102 7103 7104 7105
 EXPOSE 7201 7202 7203 7204 7205
 
-ARG PDO_HOSTNAME
+ARG PDO_HOSTNAME=localhost
 ENV PDO_HOSTNAME=$PDO_HOSTNAME
 
-ARG PDO_LEDGER_URL
-ENV PDO_LEDGER_URL=$PDO_LEDGER_URL
```
Comment (bvavala) [inline docker/pdo_services.dockerfile:62]: (same comment as above)

File: docker/services_base.yaml [hunk 6]
```diff
@@ -17,18 +17,9 @@ version: "3.4"
 
 services:
   services_container:
-    image: pdo_services
-    build:
```
Comment (bvavala) [inline docker/services_base.yaml:21]: +1 to remove this and consolidate the build
(the other way around of using just docker-compose worked as well)

**Review by bvavala — COMMENTED** [review]
Looks good overall, just one comment.

File: docker/pdo_client.dockerfile (hunk 4 above)
Comment (g2flyer) [reply docker/pdo_client.dockerfile:72]: I see your point. I think a valid PDO_HOSTNAME var s required by some of our build scripts. That said, in that case we could (should) drop the ARG part but still retain the ENV so the build works?

File: docker/pdo_client.dockerfile (hunk 1 above)
Comment (cmickeyb) [reply docker/pdo_client.dockerfile:39]: Ideally... start_client.sh would be run kind of like bashrc. But there were a bunch of other usability issues. Open to suggestions.

File: docker/pdo_client.dockerfile (hunk 4 above)
Comment (cmickeyb) [reply docker/pdo_client.dockerfile]: I'm writing up a more complete version of what needs to happen with docker. will push in an issue for discussion. We (mea culpa) are still mixing up 1) a generic image that anyone could deploy on any network and 2) an image pre-configured for a specific network. I think we should make ALL images that we build preconfigured with hostname as localhost and ledger as 127.0.0.1:6600. Then use the start_* scripts to set what actually needs to be done. Keys are already provided along similar lines.

File: docker/pdo_services.dockerfile (hunk 5 above)
Comment (cmickeyb) [reply docker/pdo_services.dockerfile]: as i mentioned above... the only good reason for having these in the dockerfile is that they simplify construction of pre-configured images. but... i'm with you... pdo-hostname and pdo-ledger-url should NOT be build variables. we should pre-configure for localhost and actually set the values we want in the start_* scripts.

File: docker/services_base.yaml (hunk 6 above)
Comment (cmickeyb) [reply docker/services_base.yaml:21]: again. I dont think this is a solution.
or maybe its a solution that we'll end up changing again in a couple weeks.

File: docker/Makefile [hunk 7]
```diff
@@ -136,14 +141,14 @@ TEST_FILES += -f services_base.yaml
...
+
+test : clean_config clean_repository test_no_reset
+
+test_no_reset : build_test stop_all
+	PDO_VERSION=$(PDO_VERSION) docker-compose $(TEST_FILES) up --abort-on-container-exit
+	PDO_VERSION=$(PDO_VERSION) docker-compose $(TEST_FILES) down
 
-test : clean_config clean_repository build_test
-	PDO_VERSION=${PDO_VERSION} docker-compose $(TEST_FILES) up --abort-on-container-exit
-	PDO_VERSION=${PDO_VERSION} docker-compose $(TEST_FILES) down
```
Comment (cmickeyb) [inline docker/Makefile:152]: this is lovely in theory... but not particularly useful in practice. if i'm re-running the tests its because i made changes to the repository. so... does this really add more value than the complexity it adds?

File: docker/base.yaml [hunk 8]
```diff
@@ -18,9 +18,6 @@ version: "3.4"
...
     image: pdo_base:${PDO_VERSION:-latest}
-    build:
-      context: .
-      dockerfile: pdo_base.dockerfile
-      args:
-        ADD_APT_PKGS: ${ADD_APT_PKGS}
-        PDO_VERSION: ${PDO_VERSION:-latest}
+    # build containers only via Makefile and directly via docker build
+    # as we use containers also outside of docker-compose to prevent
+    # double-build
```
Comment (cmickeyb) [inline docker/base.yaml:23]: ah... way ahead of me! see my comment about add SGX_MODE to the versions & i think this will help a LOT.

File: docker/pdo_ccf.dockerfile [hunk 9]
```diff
@@ -20,9 +20,9 @@ FROM pdo_ccf_base:${PDO_VERSION}
 # -----------------------------------------------------------------
 # set up the PDO sources
 # -----------------------------------------------------------------
-ARG REBUILD 0
+ARG REBUILD=0
```
Comment (cmickeyb) [inline docker/pdo_ccf.dockerfile:24]: ok.. for consistency i'm fine with this. but... turns out that "=" is optional.

File: docker/pdo_ccf.dockerfile [hunk 10]
```diff
@@ -49,11 +49,8 @@ RUN /project/pdo/tools/build_ccf.sh
...
-ARG PDO_HOSTNAME
-ENV PDO_HOSTNAME=$PDO_HOSTNAME
-
-ARG PDO_LEDGER_URL
-ENV PDO_LEDGER_URL=$PDO_LEDGER_URL
+ARG PDO_HOSTNAME=localhost
+ENV PDO_HOSTNAME=${PDO_HOSTNAME}
 
 # Note that the entry point when specified with exec syntax
 # can be extended through the docker run interface far more
```
Comment (cmickeyb) [inline docker/pdo_ccf.dockerfile:52]: this needs more discussion. adding the hostname and ledger url allows construction of pre-configured images (that is, the image contains the url). in practice, however, its WAY easier to just specify the interface and ledger in the startup script parameters.

File: docker/pdo_client.dockerfile [hunk 11]
```diff
@@ -68,16 +65,13 @@ COPY --chown=${UNAME}:${UNAME} tools/*.sh ./
...
-
-# the client is set up for interactive access; the environment can be
-# set up by source /project/pdo/tools/start_client.sh with the arguments
-# to build a new client environment or copy one from the xfer directory
+RUN \
+  echo 'echo "\
+NOTE: to setup environment, \"source /project/pdo/tools/start_client.sh\"\n\
+   with the arguments \"--mode build\" to build a new client environment (default mode)\n\
+   or \"--mode copy\" to copy one from the xfer directory"\
+' > /project/pdo/.bashrc
```
Comment (cmickeyb) [inline docker/pdo_client.dockerfile:73]: just create a .bashrc file & put this in the file. i will not approve changes that involve creating scripts in the dockerfile.

File: docker/pdo_services.dockerfile (hunk 5 above)
Comment (cmickeyb) [reply docker/pdo_services.dockerfile]: if you are keeping hostname then you have to keep the ledger_url. there is no reason whatsoever to have one without the other. either drop both or keep both.

**Review by cmickeyb — CHANGES_REQUESTED** [review]

File: docker/tools/build_client.sh [hunk 12]
```diff
@@ -15,9 +15,9 @@
 
 source /project/pdo/tools/environment.sh
 
-# these variables should be unused during build
-export PDO_HOSTNAME=
-export PDO_LEDGER_URL=
+# to get build without (ignored) errors
+export PDO_HOSTNAME=localhost
+export PDO_LEDGER_URL=https://127.0.0.1:6600
```
Comment (cmickeyb) [inline docker/tools/build_client.sh:21]: still suggest we pull these changes into a separate PR. we should not have to set these until we run.
also... push these into environment.sh. not here.

File: docker/tools/build_services.sh [hunk 13]
```diff
@@ -13,9 +13,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-# these variables should be unused during build
-export PDO_HOSTNAME=
-export PDO_LEDGER_URL=
+# to get build without (ignored) errors
+export PDO_HOSTNAME=localhost
+export PDO_LEDGER_URL=https://127.0.0.1:6600
```
Comment (cmickeyb) [inline docker/tools/build_services.sh:18]: same

File: bin/lib/common.sh [hunk 14]
```diff
@@ -82,10 +82,13 @@ function force_to_ip()
...
+function check_pdo_env()
+{
+    : "${PDO_HOME:-$(die Missing environment variable PDO_HOME)}"
+    : "${PDO_HOSTNAME:-$(die Missing environment variable PDO_HOSTNAME)}"
+    : "${PDO_INTERPRETER:-$(die Missing environment variable PDO_INTERPRETER)}"
+    : "${PDO_LEDGER_KEY_ROOT:-$(die Missing environment variable PDO_LEDGER_KEY_ROOT)}"
+    : "${PDO_LEDGER_TYPE:-$(die Missing environment variable PDO_LEDGER_TYPE)}"
+    : "${PDO_LEDGER_URL:-$(die Missing environment variable PDO_LEDGER_URL)}"
+    : "${PDO_SOURCE_ROOT:-$(die Missing environment variable PDO_SOURCE_ROOT)}"
+}
```
[the remark this answers is not in the document]
Comment (g2flyer) [reply bin/lib/common.sh:95]: It actually does, as otherwise there were error (messages) during build ....

File: bin/lib/common.sh [hunk 15]
```diff
@@ -47,7 +47,7 @@ function yell () {
 }
 
 function die() {
-    recho "$(basename $0): $*" >&2
+    recho "$(basename $0): ERROR: $*" >&2
     exit 111
```
Comment (cmickeyb) [inline bin/lib/common.sh:51]: not to be pedantic... but did you check the use of this in the MANY other places its used? is "error:" really necessary?

File: docker/Makefile [hunk 16]
```diff
@@ -136,14 +141,14 @@ TEST_FILES += -f services_base.yaml
 TEST_FILES += -f ccf_base.yaml
 TEST_FILES += -f test.yaml
 
-build_test : repository build_base build_services_base build_ccf_base
-	PDO_VERSION=${PDO_VERSION} PDO_USER_UID=$(PDO_USER_UID) PDO_GROUP_UID=$(PDO_GROUP_UID) \
-		docker-compose $(TEST_FILES) build \
-			--build-arg PDO_VERSION=${PDO_VERSION}
+build_test : repository build_services build_ccf build_client
+
```
[the remark this answers is not in the document]
Comment (g2flyer) [reply docker/Makefile:145]: this is already done? There shouldn't be any build section in the compose-files anymore but we of course still needed the build targets in the makefile to build using docker build?

File: docker/Makefile (hunk 7 above)
Comment (g2flyer) [reply docker/Makefile:152]: it tremendously helps, e.g., if you change anything related to docker building which is _not_ taking from the repository (which of course arguably could be seen as an issue)

File: docker/base.yaml (hunk 8 above)
Comment (g2flyer) [reply docker/base.yaml:23]: see above, i don't think adding sgx-mode to tag is the right thing to do

File: docker/pdo_ccf.dockerfile (hunk 9 above)
Comment (g2flyer) [reply docker/pdo_ccf.dockerfile:24]: i know it's optional, just did it for consistency ...

File: docker/pdo_ccf.dockerfile (hunk 10 above)
Comment (g2flyer) [reply docker/pdo_ccf.dockerfile:52]: this is an outdated change regardless of github not marking it as such. In fact, the actual change is exactly the opposite (i.e., exactly what you want?)

File: docker/pdo_client.dockerfile (hunk 11 above)
Comment (g2flyer) [reply docker/pdo_client.dockerfile:73]: yeah, i was thinking about that but then didn't really like to put a .bashrc file, so i decided on this.  But then in copy i can change file name, so while but it as pdo_client.bashrc  .and then COPY it ...

File: docker/Makefile [hunk 17]
```diff
@@ -60,36 +58,43 @@ all : $(addprefix build_,$(IMAGES))
...
-	@ docker rmi -f pdo_$*:${PDO_VERSION}
+	@ docker rmi -f pdo_$*:$(PDO_VERSION)
 
 run_ccf : build_ccf stop_ccf
-	@ docker run -v $(DOCKER_DIR)/xfer/:/project/pdo/xfer --network host --name ccf_container -P -d pdo_ccf:${PDO_VERSION}
+	@ docker run -v $(DOCKER_DIR)/xfer/:/project/pdo/xfer --network host --name ccf_container -P -d pdo_ccf:$(PDO_VERSION)
 
 run_ccf_dev : build_ccf
-	@ docker run -v $(DOCKER_DIR)/xfer/:/project/pdo/xfer -v ${PDO_SOURCE_ROOT}/docker/tools/:/project/pdo/tools -v ${PDO_SOURCE_ROOT}/:/project/pdo/src --entrypoint=/bin/bash -it --network host --name ccf_container_dev -P pdo_ccf:${PDO_VERSION} -c "stty cols $$(tput cols) rows $$(tput lines) && bash"
+	@ docker run -v $(DOCKER_DIR)/xfer/:/project/pdo/xfer -v ${PDO_SOURCE_ROOT}/docker/tools/:/project/pdo/tools -v ${PDO_SOURCE_ROOT}/:/project/pdo/src --entrypoint=/bin/bash -it --network host --name ccf_container_dev -P pdo_ccf:$(PDO_VERSION) -c "stty cols $$(tput cols) rows $$(tput lines) && bash"
```
[the remark this answers is not in the document]
Comment (g2flyer) [reply docker/Makefile:86]: that is true once we go to a registry and unlike windows and mac seems docker on linux doesn't offer some convenience uid/gid mapping.  That said, isn't this also related what i mentioned above as question in the PS: whether we should follow also here the naming convention of including user into image names so we can also safely develop on multi-user machine?  In which case the _dev versions would also work fine.  Is of course also related to the below naming convention regarding sgx-mode and alike (and indirectly what our strategy would be regarding "production" setups and related to that how to manage/organise the regstiry?

File: bin/lib/common.sh (hunk 15 above)
Comment (g2flyer) [reply bin/lib/common.sh:51]: well,  we do an exit afterwards and so suggests the function name. So clearly it is considered an error (except of course the exit is ignored on some places where we do `: good-condition || $(die ....)` that this is clearly still semantically an error.
I actulaly also add a WARN to yell but we seem to use yell also for non-warning cases so i dropped that.  The main reason i added that is that i wanted to be able to grep for these messages in the voluminous log-fiels .  I guess there is a way to also grep for the color red but i think these seemed cleaner and works even on b/w terminals ...

File: docker/pdo_client.dockerfile (hunk 11 above)
Comment (g2flyer) [reply docker/pdo_client.dockerfile]: fixed ...

File: docker/pdo_client.dockerfile (hunk 4 above)
Comment (cmickeyb) [reply docker/pdo_client.dockerfile]: > I thoutht i could split check_pdo_env in runtime and build-time check but alas with some tests scripts during build it willl also be required. 

which scripts do you think we run at build time? specifically for docker...

File: docker/Makefile (hunk 17 above)
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docker/Makefile:86]: suggest you move all the dev tags to a separate make file that can be included. and include LOTS of comments about the limitations of these targets. They are very useful... but the problems are not intuitive.

File: docker/Makefile (hunk 7 above)
Comment (cmickeyb) [reply docker/Makefile:152]: but how often to you do this? that's a very small fraction of the time.

File: docker/Makefile [hunk 18]
```diff
@@ -60,36 +58,43 @@ all : $(addprefix build_,$(IMAGES))
 rebuild_% : repository
 	@ docker build $(DOCKER_ARGS) \
 		--build-arg REBUILD=$(TIMESTAMP) \
-		--build-arg PDO_VERSION=${PDO_VERSION} \
-		--tag pdo_$*:${PDO_VERSION} \
+		--build-arg PDO_VERSION=$(PDO_VERSION) \
+		--tag pdo_$*:$(PDO_VERSION) \
 		--file $(DOCKER_DIR)/pdo_$*.dockerfile .
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docker/Makefile:63]: naming might be easier with tags, but i agree that the best way to do this is to push it into the container name itself. good suggestion!

File: docker/pdo_client.bashrc [hunk 19]
```diff
@@ -0,0 +1,5 @@
+/usr/bin/echo -e "\
+NOTE: to setup environment, \"source /project/pdo/tools/start_client.sh\"\n\
+   with the arguments \"--mode build\" to build a new client environment (default mode)\n\
+   or \"--mode copy\" to copy one from the xfer directory"
+
```
Comment (cmickeyb) [inline docker/pdo_client.bashrc:5]: put this in the tools directory.

Comment (g2flyer) [reply docker/pdo_client.bashrc:5]: i wouldn't have seen this really as a tool and we still have to do dockerfile-action as the copying of tools is not enough?  But if you prefer i can put it there and then replace the current COPY with with RUN 'ln -s ...' command ...

**Comment by g2flyer** [comment]
@cmickeyb the latest version should address your comments and merge in the changes from your branch, with slight modifications for make.dev ...

**Review by cmickeyb — APPROVED** [review]

**Review by bvavala — APPROVED** [review]
looks good
